patches/0005-Prefer-Quilt-over-Fabric-Loader.patch

```diff
@@ -5,66 +5,46 @@ Subject: [PATCH] Prefer Quilt over Fabric Loader
  import com.google.gson.Gson;
  import com.google.gson.GsonBuilder;
 +
 +import net.fabricmc.loom.util.FabricCapabilities;
 +
  import org.gradle.api.Project;
++import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
  import org.gradle.api.plugins.PluginAware;
  
-@@ -83,6 +86,9 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
+ import net.fabricmc.loom.api.LoomGradleExtensionAPI;
+@@ -83,6 +87,18 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
  		project.apply(ImmutableMap.of("plugin", "java-library"));
  		project.apply(ImmutableMap.of("plugin", "eclipse"));
  
 +		// Tell Gradle that Quilt Loader/Fabric Loader are the same thing
 +		project.getDependencies().getComponents().all(FabricCapabilities.class);
-+
- 		// Setup extensions
- 		project.getExtensions().create(LoomGradleExtensionAPI.class, "loom", LoomGradleExtensionImpl.class, project, LoomFiles.create(project));
- 		project.getExtensions().create("fabricApi", FabricApiExtension.class);
-diff --git a/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java b/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
-index 009d91ecafc42640135d9ebf5c1eb06be323e115..ac9c0ac012c82a50cb95eb4a9ac8a77613ae255f 100644
---- a/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
-+++ b/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
-@@ -43,6 +43,7 @@ import org.gradle.api.artifacts.Configuration;
- import org.gradle.api.artifacts.FileCollectionDependency;
- import org.gradle.api.artifacts.MutableVersionConstraint;
- import org.gradle.api.artifacts.ResolvedArtifact;
-+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
- import org.gradle.api.artifacts.dsl.DependencyHandler;
- import org.gradle.api.artifacts.query.ArtifactResolutionQuery;
- import org.gradle.api.artifacts.result.ArtifactResult;
-@@ -131,6 +132,18 @@ public class ModConfigurationRemapper {
- 				configsToRemap.put(entry.getSourceConfiguration().get(), remappedConfig);
- 			}
- 		}
-+		// Round 0: Constraints
-+		// TODO: this is probably the wrong place for this
-+		// Explain to Gradle how to handle conflicts
-+		for (Configuration c : configsToRemap.keySet()) {
++		project.getConfigurations().configureEach(c -> {
 +			// this mirrors how it's done at https://docs.gradle.org/current/userguide/dependency_capability_conflict.html#sub:capabilities for gradle 8.10.2
 +			var rs = c.getResolutionStrategy().getCapabilitiesResolution();
 +			rs.withCapability("net.fabricmc:fabric-loader", a -> {
-+				a.getCandidates().stream().filter(id -> id instanceof ModuleComponentIdentifier && ((ModuleComponentIdentifier) id).getModule().equals("quilt-loader"))
++				a.getCandidates().stream().filter(id -> id.getId() instanceof ModuleComponentIdentifier moduleId && moduleId.getModule().equals("quilt-loader"))
 +						.findFirst().ifPresent(a::select);
```

Comment:
Couldn't this cause an _older_ version of qloader to be selected?
But either way, this should handle the case when the version of the transitive floader dependency is newer than the floader version the newest qloader version in the list provides, failing the build in that case. If the user wants to ignore the transitive dependency's requests, that should be done explicitly using an exclude, not implicitly.
I think once the TODO mentioned in the other review is fixed, just using `selectHighestVersion` _should_ do the trick.

Reply:
Yes, but like we have talked, you *want* the user-chosen Quilt Loader to supercede *all* transitives Quilt Loaders, although this is on the implicit exclusion zone; Maybe your approach would be technically better (and you have presented decent solutions like [the GradleX plugin](https://github.com/gradlex-org/jvm-dependency-conflict-resolution)), but I don't see reasons why we should be more hostile to the user in the process when other toolchains have just completely hidden the problem on the first place;

While I'd maybe be fine in an alternate universe where Quilt was more dominant (mainly because mods would be more willing to fix their deps)? We don't live there; right now, we are burdened by an ecosystem that does not want to adapt towards us; so yeah.. explicit excluding (and encouragement to not have transient deps) won't happen as far as I know

Reply:
We didn't talk about that part specifically, no. Without any customization, the standard Gradle mechanisms do exactly that: use the latest version of the candidates that fulfills all requirements. It doesn't care about which of those the user specifies. (well, perhaps that should also be a failure scenario... buuut yeah)
Maven prefers what the user specifies... by preferring whatever is closest to the root of resolution, IIRC, completely disregarding everything else like a dependency deeper down wanting Foo 9.6 instead of Foo 1.2 (it should be obvious by now that Maven is not a good source of ideas)
So imagine this: I have a Fabric project depending on Loader 0.6 or something, and add a dependency on one depending on 0.15 and actually using newly added APIs. Gradle will silently resolve 0.15, and it'll work (although perhaps not as intended).
I now have a Quilt project depending on Quilt Loader 0.16 (which is ofc older) and add a dependency on that peoject. As implemented, it'll silently **ignore** the Fabric Loader dependency, and silently use Quilt Loader 0.16 - which will break.
Ignoring dependencies like that should always be conscious and explicit decisions by the user.